src/gateway/Gateway.sol

```diff
@@ -51,47 +73,277 @@ contract Gateway is Auth, IGateway {
+            require(activeRouters[router].id != 0, "Gateway/invalid-router");
+            recoveries[messageHash] = Recovery(block.timestamp + RECOVERY_CHALLENGE_PERIOD, router);
+            emit InitiateMessageRecovery(messageHash, router);
+        } else if (MessagesLib.messageType(payload) == MessagesLib.Call.DisputeMessageRecovery) {
+            bytes32 messageHash = payload.toBytes32(1);
+            return _disputeMessageRecovery(messageHash);
+        }
+    }
+
+    /// @inheritdoc IGateway
+    function disputeMessageRecovery(bytes32 messageHash) external auth {
+        _disputeMessageRecovery(messageHash);
+    }
+
+    function _disputeMessageRecovery(bytes32 messageHash) internal {
+        delete recoveries[messageHash];
+        emit DisputeMessageRecovery(messageHash);
+    }
+
+    /// @inheritdoc IGateway
+    function executeMessageRecovery(bytes calldata message) external {
+        bytes32 messageHash = keccak256(message);
+        // wouldn't it better to mark these as memory?
+        Recovery storage recovery = recoveries[messageHash];
+        Router storage router = activeRouters[recovery.router];
+
+        require(recovery.timestamp != 0, "Gateway/message-recovery-not-initiated");
+        require(recovery.timestamp <= block.timestamp, "Gateway/challenge-period-has-not-ended");
+        require(router.id != 0, "Gateway/invalid-router");
+
+        delete recoveries[messageHash];
+        _handle(message, recovery.router, router, true);
+        emit ExecuteMessageRecovery(message);
+    }
+
     // --- Outgoing ---
-    function send(bytes calldata message) public pauseable {
+    /// @inheritdoc IGateway
+    function send(bytes calldata message, address source) public payable pauseable {
         require(
-            msg.sender == investmentManager || msg.sender == poolManager || msg.sender == messages[message.toUint8(0)],
+            msg.sender == investmentManager || msg.sender == poolManager
+                || msg.sender == messageHandlers[message.toUint8(0)],
             "Gateway/invalid-manager"
         );
-        aggregator.send(message);
+
+        bytes memory proof = abi.encodePacked(uint8(MessagesLib.Call.MessageProof), keccak256(message));
+
+        uint256 numRouters = routers.length;
+        require(numRouters > 0, "Gateway/routers-not-initialized");
+        uint256 fuel = quota;
+        uint256 tank = fuel > 0 ? fuel : gasService.shouldRefuel(source, message) ? address(this).balance : 0;
+
+        uint256 messageCost = gasService.estimate(message);
+        uint256 proofCost = gasService.estimate(proof);
+
+        for (uint256 i; i < numRouters; i++) {
+            RouterLike currentRouter = RouterLike(routers[i]);
+            bool isPrimaryRouter = i == PRIMARY_ROUTER_ID - 1;
+            bytes memory payload = isPrimaryRouter ? message : proof;
+
+            uint256 consumed = currentRouter.estimate(payload, isPrimaryRouter ? messageCost : proofCost);
+            uint256 remaining;
```

Comment:
I would not introduce `remaining` as a new var. You are risking an overflow, as tank might be smaller than consumed. Would need to use an int. But also the var is not required. Would do following:

```
   require(consumed < tank, "Gateway/not-enough-gas-funds");
   tank = tank - consumed;
```

Reply:
Actually I was relying on the underflow. That's why there was this `unchecked`. But I agree that we don't need the remaining. 
Cheers.